aplib.net-demo/Assets/Testing/AplibTests/ConnectedComponentsTests.cs

```diff
@@ -1,35 +1,84 @@
 using Aplib.Core;
-using Aplib.Core.Agents;
 using Aplib.Core.Belief.Beliefs;
 using Aplib.Core.Belief.BeliefSets;
-using Aplib.Core.Desire.DesireSets;
-using Aplib.Core.Desire.Goals;
-using Aplib.Core.Desire.GoalStructures;
-using Aplib.Core.Intent.Actions;
-using Aplib.Core.Intent.Tactics;
 using Aplib.Integrations.Unity;
-using Aplib.Integrations.Unity.Actions;
 using LevelGeneration;
 using NUnit.Framework;
 using System.Collections;
+using System.Collections.Generic;
 using System.Linq;
 using UnityEngine;
 using UnityEngine.SceneManagement;
 using UnityEngine.TestTools;
+using static Aplib.Core.Combinators;
+using Action = Aplib.Core.Intent.Actions.Action<Testing.AplibTests.ConnectedComponentsBeliefSet>;
+using BdiAgent = Aplib.Core.Agents.BdiAgent<Testing.AplibTests.ConnectedComponentsBeliefSet>;
+using DesireSet = Aplib.Core.Desire.DesireSets.DesireSet<Testing.AplibTests.ConnectedComponentsBeliefSet>;
+using Goal = Aplib.Core.Desire.Goals.Goal<Testing.AplibTests.ConnectedComponentsBeliefSet>;
+using PrimitiveTactic = Aplib.Core.Intent.Tactics.PrimitiveTactic<Testing.AplibTests.ConnectedComponentsBeliefSet>;
+using Tactic = Aplib.Core.Intent.Tactics.Tactic<Testing.AplibTests.ConnectedComponentsBeliefSet>;
+using TransformPathfinderAction = Aplib.Integrations.Unity.Actions.TransformPathfinderAction<Testing.AplibTests.ConnectedComponentsBeliefSet>;
 
 namespace Testing.AplibTests
 {
-    public class ConnectedComponentsBeliefSet : BeliefSet
+    public class ConnectedComponentsBeliefSet : IBeliefSet
     {
-        public Belief<GameObject, Rigidbody> PlayerRigidbody = new(
-            GameObject.Find("Player"),
-            x => x.GetComponent<Rigidbody>());
+        public Rigidbody PlayerRigidbody { get; } = GameObject.Find("Player").GetComponent<Rigidbody>();
+
+        public Belief<Queue<Vector3>, Vector3> GetCurrentTarget;
+
+        public Queue<Vector3> TargetPositionsInConnectedComponents { get; } = new(GetTargetPositionsInConnectedComponents());
+
+        public Queue<Vector3> TeleporterPositions { get; } = new(GetTeleporterLandingPoints());
+
+        public ConnectedComponentsBeliefSet()
+        {
+            GetCurrentTarget = new(
+                reference: TargetPositionsInConnectedComponents,
+                getObservationFromReference: positions => positions.Peek(),
+                shouldUpdate: () =>
+                {
+                    if (TargetPositionsInConnectedComponents.Count == 0) return false;
```

Comment:
Seems like this covers an edge case. Can you prevent this edge case from happening at all?

Reply:
No, hahaha. The shouldUpdate lambda is called every frame, but eventually there will be no TargetPositions left, and then we cannot Peek, because that gives an exception.